jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -781,10 +781,22 @@ static MethodHandle adaptInterfaceLookupsOfObjectMethodsIfRequired(MethodHandle
 		/*
 		 * Check for methods MethodHandle.invokeExact, MethodHandle.invoke,
 		 * or MethodHandles.varHandleInvoker.  
-		 * Access checks are not required as these methods are public and therefore
+		 * Access checks to these methods are not required as these methods are public and therefore
 		 * accessible to everyone.
+		 * However the access checks to the parameter and return classes are required.
 		 */
-		static MethodHandle handleForMHInvokeMethods(Class<?> clazz, String methodName, MethodType type) {
+		MethodHandle handleForMHInvokeMethods(Class<?> clazz, String methodName, MethodType type) throws IllegalAccessException {
+			if (INTERNAL_PRIVILEGED != accessMode) {
```

Comment:
We're likely doing double access checks as this check is being done even if that the method being looked up isn't MH.invoke/invokeExact, or VarHandle.

Can this be modified to only do the check if the MH is for invoke/invokeExact or VarHandle invoker?

Reply:
Added a helper method to perform `parameter/return` class access check for `invoke/invokeExact` and `VarHandle invoker`.
Verified manually that `pConfig` still compiles.
@DanHeidinga please have another look.